File: mod10/main.go
package main

import (
        "bufio"
        "fmt"
        "os"
        "strconv"
        "strings"
)

func findLongestChainEndpoint(pairs [][]string) string {
        if len(pairs) == 0 {
                return "Нет пар для анализа"
        }

        graph := make(map[string]string)
        for _, pair := range pairs {
                if pair[0] == pair[1] {
                        return "обратитесь к специалисту"
                }
                graph[pair[0]] = pair[1]
        }

        longestChainEndpoint := ""
        maxLength := 0

        for start := range graph {
                current := start
                length := 1
                for {
                        next, ok := graph[current]
                        if !ok {
                                break
                        }
                        current = next
                        length++
                }

                if length > maxLength {
                        maxLength = length
                        longestChainEndpoint = current
                }
        }

        return longestChainEndpoint
}

func main() {
        reader := bufio.NewReader(os.Stdin)

        fmt.Println("Введите количество пар:")
        numPairsStr, _ := reader.ReadString('\n')
        numPairs, err := strconv.Atoi(strings.TrimSpace(numPairsStr))
        if err != nil || numPairs <= 0 {
                fmt.Println("Некорректный ввод количества пар.")
                return
        }

        pairs := make([][]string, numPairs)
        for i := 0; i < numPairs; i++ {
                fmt.Println("Введите пару (начальный пункт конечный пункт):")
                input, _ := reader.ReadString('\n')
                parts := strings.Fields(input)
                if len(parts) == 2 {
                        pairs[i] = parts
                } else {
                        fmt.Println("Некорректный ввод пары. Попробуйте снова.")
                        return
                }
        }

        // Проверка на пустые пары внутри массива.
        filteredPairs := make([][]string, 0)
        for _, pair := range pairs {
                if len(pair) == 2 && pair[0] != "" && pair[1] != "" {
                        filteredPairs = append(filteredPairs, pair)
                }
        }

        result := findLongestChainEndpoint(filteredPairs)
        fmt.Println("Итоговый конечный пункт:", result)
}

File: mod10/main.rb
def find_longest_chain_endpoint(pairs)
  if pairs.empty?
    return "Нет пар для анализа"
  end

  graph = {}
  pairs.each do |pair|
    if pair[0] == pair[1]
      return "обратитесь к специалисту"
    end
    graph[pair[0]] = pair[1]
  end

  longest_chain_endpoint = ""
  max_length = 0

  graph.keys.each do |start|
    current = start
    length = 1
    while graph.key?(current)
      current = graph[current]
      length += 1
    end

    if length > max_length
      max_length = length
      longest_chain_endpoint = current
    end
  end

  return longest_chain_endpoint
end

def main
  print "Введите количество пар:\n"
  num_pairs = gets.chomp.to_i

  if num_pairs <= 0
    puts "Некорректный ввод количества пар."
    return
  end

  pairs = []
  num_pairs.times do
    print "Введите пару (начальный пункт конечный пункт):\n"
    input = gets.chomp
    parts = input.split(" ")
    if parts.length == 2
      pairs << parts
    else
      puts "Некорректный ввод пары. Попробуйте снова."
      return
    end
  end

  # Проверка на пустые пары внутри массива.
  filtered_pairs = pairs.select { |pair| !pair[0].empty? && !pair[1].empty? }

  result = find_longest_chain_endpoint(filtered_pairs)
  puts "Итоговый конечный пункт: #{result}"
end

main